Amplify/Core/Error/CoreError.swift

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// Errors associated with operations provided by Amplify
+public enum CoreError {
+
+    /// A related operation performed on `List` resulted in an error.
+    case listOperation(ErrorDescription, RecoverySuggestion, Error? = nil)
+
+    /// A client side validation error occured.
+    case clientValidation(ErrorDescription, RecoverySuggestion, Error? = nil)
+
+    /// An error to encapsulate an error received by a dependent plugin
+    case pluginError(AmplifyError)
+
+    /// An unknown error occurred
+    case unknown(ErrorDescription, Error? = nil)
```

Comment:
Are we using this?

Reply:
we are hardcoding to `unknown` in DataStoreListProvider's synchronous `load()` as the default case before using a DispatchSemaphore on the async load. So it will never occur and will hang on the sempahore if the query doesn't return. I have a few options for this, most likey i can do `.pluginError(DataStore.operationError)` or some sort as the default 

One reason was was because `AmplifyError` requires an init, which we've implemented across other AmplifyError's default to setting it to `.unknown` in the else case:
```
public init(
        errorDescription: ErrorDescription = "An unknown error occurred",
        recoverySuggestion: RecoverySuggestion = "(Ignored)",
        error: Error
    ) {
        if let error = error as? Self {
            self = error
        } else {
            self = .unknown(errorDescription, error)
        }
    }
```

`.unknown` does have a slight difference from the others, for example, we will return recoverySuggestion as `AmplifyErrorMessages.shouldNotHappenReportBugToAWS()` which will guide developers to open a bug with the issue that occured:
```
There is a possibility that there is a bug if this error persists. Please take a look at \
        https://github.com/aws-amplify/amplify-ios/issues to see if there are any existing issues that \
        match your scenario, and file an issue with the details of the bug if there isn't. Issue encountered \
        at:
```